### PR TITLE
Opt/pretraining merge main

### PR DESCRIPTION
**Pull request by shrrosier**
Merge of opt/pretraining which includes:
- entirely new pretraining module 
- alternative data assimilation module (data_assimilation_SR) which is a WIP

The only change affecting the main iceflow pipeline is that the network normalizer is treated differently depending on if you use a pretrained network or not.

All unit tests passed.

### CODE REVIEW

File: igm/processes/iceflow/emulate/utils/normalizations.py [hunk 1]
```diff
@@ -334,3 +334,40 @@ def get_config(self):
             }
         )
         return config
+
+
+class FixedChannelStandardization(tf.keras.layers.Layer):
```
Comment (brfi3983) [inline igm/processes/iceflow/emulate/utils/normalizations.py:339]: Is this not the same functionality as the `FixedAffine` normalization layer?

I believe both should be channel wise :)

File: igm/processes/iceflow/energy/components/sliding/laws/weertman.py [hunk 2]
```diff
@@ -133,9 +137,13 @@ def _cost(
     # Effective exponent
     p = 1.0 + 1.0 / m
 
+    C_h = tau_ref_h / tf.pow(u_ref, 1.0 / m)
```
Comment (brfi3983) [inline igm/processes/iceflow/energy/components/sliding/laws/weertman.py:140]: Great job adding the new sliding law - should simplify dealing with the units great :D Could you also do the same for `budd.py` and `coulomb.py` to maintain consistency between all the sliding laws?

File: igm/processes/iceflow/unified/bcs/no_inflow.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    """Open boundary: outflow allowed, inflow suppressed (normal component only)."""
+
+    def apply(self, U: TV, V: TV) -> Tuple[TV, TV]:
+        # --- 1) copy interior values to the boundary (Neumann-ish) to avoid a "dam"
+        # y boundaries
+        U = tf.concat([U[:, :, 1:2, :], U[:, :, 1:-1, :], U[:, :, -2:-1, :]], axis=2)
+        V = tf.concat([V[:, :, 1:2, :], V[:, :, 1:-1, :], V[:, :, -2:-1, :]], axis=2)
+        # x boundaries
+        U = tf.concat([U[:, :, :, 1:2], U[:, :, :, 1:-1], U[:, :, :, -2:-1]], axis=3)
+        V = tf.concat([V[:, :, :, 1:2], V[:, :, :, 1:-1], V[:, :, :, -2:-1]], axis=3)
```
Comment (brfi3983) [inline igm/processes/iceflow/unified/bcs/no_inflow.py:21]: @tgregov Can you check if this consistent with all the horizontal discretizations (i.e. `P1` and `Q1`) or if this will this only work with finite differences?

File: igm/processes/iceflow/unified/mappings/transforms.py [hunk 4]
```diff
@@ -79,22 +98,44 @@ def theta_bounds(
...
+    def to_physical(self, theta: tf.Tensor) -> tf.Tensor:
+        return theta / tf.cast(1000.0, theta.dtype)
+
+    def theta_bounds(self, lower_phys, upper_phys, dtype, eps: float = 1e-12):
+        L = -tf.constant(float("inf"), dtype) if lower_phys is None else tf.constant(lower_phys * 1000.0, dtype)
+        U =  tf.constant(float("inf"), dtype) if upper_phys is None else tf.constant(upper_phys * 1000.0, dtype)
+        return L, U
+    
 # registry
 TRANSFORMS: Dict[str, Type[ParameterTransform]] = {
```
Comment (brfi3983) [inline igm/processes/iceflow/unified/mappings/transforms.py:136]:
```suggestion
Transforms: Dict[str, Type[ParameterTransform]] = {
```

Just a nitpick, but we should try to be consistent with the naming conventions found in other files. If you can make the other changes with the naming, that would be great :)

File: igm/processes/iceflow/unified/solver/solver.py [hunk 5]
```diff
@@ -78,8 +78,10 @@ def solve_iceflow(cfg: DictConfig, state: State, init: bool = False) -> None:
     if do_solve:
         inputs = get_solver_inputs_from_state(cfg, state)
 
-        # Compute and set stats can be linked inside the class, so we can get rid of the liskov violation here (when means in one transformation is not means in another)
-        means, variances = mapping.input_normalizer.compute_stats(inputs)
-        mapping.input_normalizer.set_stats(means, variances)
+        if not cfg.processes.iceflow.unified.network.pretrained: # the pretrained network uses a different normalization scheme
```
Comment (brfi3983) [inline igm/processes/iceflow/unified/solver/solver.py]: I understand that the data assimilation uses an initial normalization based on the training dataset; however, if afterwards we are stepping forward in time, don't we still need to re-normalize our `fieldin` using our forward model (and not just the pretrained stats)?

If this is true, could we maybe try to
1. Check if we are using pretraining, and if so, change the `status` to be pretrained.
2. Use this to determine the initial stats (i.e. you can just call the `input_normalizer.set_stats` and read it from your emulator's training database stats.
3. Then, after the inversion, the status will change from `pretrained` and retrain/minimize normally for the forward time stepping.

File: igm/conf/processes/data_assimilation_SR.yaml [hunk 6]
```diff
@@ -0,0 +1,13 @@
+data_assimilation_SR:
```
Comment (brfi3983) [inline igm/conf/processes/data_assimilation_SR.yaml:1]: Should we try to think of a dedicated name instead of `_SR` so that it is a bit clear what your method is doing? This way we do not have three different data assimilation modules like `_KL`, _`SK`, etc.

File: igm/conf_help/processes/pretraining.yaml
Comment (brfi3983) [inline igm/conf_help/processes/pretraining.yaml:1]: So nice to see his file cleaned up 😃

File: igm/processes/data_assimilation_SR/outputs/output_ncdf.py
Comment (brfi3983) [inline igm/processes/data_assimilation_SR/outputs/output_ncdf.py:1]: Could we move this file to the default `igm/outputs/` folder and then call it directly? Or will it break the code ? If we do this, we can alter the `igm_run.py` script so that during data assimilation, it can call it instead of having to wait for the end of the time loop on each iteration.

Separately, what are your thoughts on having the naming convention more like `data_assimilation` where it would ideally work for any of the DA methods (KL, GJ, etc.)?

File: igm/processes/data_assimilation_SR/terms.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            dx=ctx.dx,
+            lam=lam,
+            mask=mask,
+            A_domain=tf.cast(ctx.A_domain, dtype),
+            eps=float(self.spec.eps),
+            ref=ref_tensor,
+        )
+
+
+MISFIT_REGISTRY = {
```
Comment (brfi3983) [inline igm/processes/data_assimilation_SR/terms.py]:
```suggestion
MisfitRegistry = {
```

For consistency with other variable names. We can decide if we like one convention more - it is certainty not set it stone at the moment.

File: igm/processes/iceflow/unified/optimizers/lbfgs.py [hunk 8]
```diff
@@ -98,6 +96,10 @@ def _compute_direction(
         last_y = y_list[num_elems - 1]
         last_s = s_list[num_elems - 1]
         gamma = self._dot(last_y, last_s) / (self._dot(last_y, last_y) + self.eps)
+
+        # basic safety checks
+        gamma = tf.where(tf.math.is_finite(gamma), gamma, tf.constant(1.0, gamma.dtype))
+        gamma = tf.clip_by_value(gamma, tf.constant(1e-6, gamma.dtype), tf.constant(1e6, gamma.dtype))
```
Comment (brfi3983) [inline igm/processes/iceflow/unified/optimizers/lbfgs.py]: @tgregov Could you check if setting fixed bounds will cause any issues depending on the scaling of the network? I can imagine a few architectures where the scaling might be numerically stable but still outside this range.

Another approach can be through the `get_grad()` method that computes `grad_u` and `grad_theta` where we can limit the gradient norms or have that as a hyperparameter so that it is not hardcoded.

File: igm/processes/pretraining/cost_tmp.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+from igm.common import State
+from igm.processes.iceflow.energy.energy import iceflow_energy_UV
+from igm.processes.iceflow.energy.utils import get_energy_components
+
+DESIRED_INPUTS = ("thk", "usurf", "arrhenius", "slidingco", "dX")
+
+"""
+This is a temporary replacement for the default energy cost since that does not currently support certain fields missing from the inputs cfg 
+(e.g. no Arrhenius-related channels, or no sliding coefficient channel). This is needed to allow for pretraining more varied networks but
+hopefully can be removed in the future once we have a more flexible + robust energy cost implementation that can handle missing fields more gracefully.
```
Comment (brfi3983) [inline igm/processes/pretraining/cost_tmp.py:17]: Hm yes, this is indeed a current limitation. There should be a way, and I have made some progress on it, but it is not ready yet :)

**Review by brfi3983 — CHANGES_REQUESTED** [review]
Thanks for all the great work - its cool to see the data assimilation module come together 😃 

I went through the file and left various comments - not all of them are rigid - which we can all discuss together the best options.

Some of the main comments:
1. I suggest we rename the `_SR` to something more general and understandable. If you can modify this extension in each file, that would be great.
2. The main question that I can see as a potential bug would be the normalization scheme. If we can make sure there are no conflicts with the current approach, that would be great. We can discuss this further with @tgregov .
3. A slight question I have is on the `no_inflow.py` boundary conditions and if they are consistent with the new horizontal discretization schemes. Again, maybe @tgregov can comment if all is good and we can move forward.
4. Lastly, we should be able to modify the `budd.py` and `coulomb.py` files with the new dimensionless sliding reference approach to keep consistent with your changes in `weertman.py`.

Thanks again for all this! We can discuss the final changes and merge it quite soon :)

File: igm/processes/iceflow/unified/optimizers/lbfgs.py (hunk 8 above)
Comment (tgregov) [reply igm/processes/iceflow/unified/optimizers/lbfgs.py]: Yep, in any case i would recommend putting these hard-coded values as arguments in the method or in the constructor of the class, with maybe some defaults values

File: igm/conf/processes/pretraining.yaml [hunk 10]
```diff
@@ -1,13 +1,9 @@
 pretraining:
-  data_dir: "surflib3d_shape_100"
+  data_dir: "/path/to/tfrecords/"
```
Comment (tgregov) [inline igm/conf/processes/pretraining.yaml:2]: Maybe avoid using a default path with `/`, as this might not work across all operating systems?

File: igm/processes/iceflow/emulate/utils/architectures/cnns.py [hunk 11]
```diff
@@ -138,18 +138,36 @@ def _build_layers(self, nb_inputs, nb_outputs):
             # Activation
             if self.activation_name.lower() == "leakyrelu":
```
Comment (tgregov) [inline igm/processes/iceflow/emulate/utils/architectures/cnns.py]: For code readability, you can use something like `activation_name = self.activation_name.lower()` before all the if statements :)

File: igm/processes/iceflow/energy/components/sliding/laws/weertman.py [hunk 12]
```diff
@@ -52,16 +52,17 @@ def cost_weertman(
 
     h = fieldin["thk"]
     s = fieldin["usurf"]
-    C = fieldin["slidingco"]
+    tau_ref = fieldin["slidingco"]
```
Comment (tgregov) [inline igm/processes/iceflow/energy/components/sliding/laws/weertman.py:55]: Can you put a comment like 
```
# TODO: rename slidingco 
```
here? To keep track of what we would need to do later on :)

File: igm/processes/iceflow/unified/optimizers/lbfgs.py [hunk 13]
```diff
@@ -125,17 +128,48 @@ def _apply_step(
         self, theta_flat: tf.Tensor, alpha: tf.Tensor, p_flat: tf.Tensor
     ) -> Tuple[tf.Tensor, Optional[tf.Tensor]]:
         return theta_flat + alpha * p_flat, None
-
+
```
Comment (tgregov) [inline igm/processes/iceflow/unified/optimizers/lbfgs.py:135]: Maybe save the file with the formatter :)

File: igm/processes/iceflow/unified/optimizers/lbfgs.py [hunk 14]
```diff
@@ -220,7 +252,7 @@ def minimize_impl(self, inputs: tf.Tensor) -> tf.Tensor:
         # Accessory variables
         halt_status = tf.constant(HaltStatus.CONTINUE.value, dtype=tf.int32)
         iter_last = tf.constant(-1, dtype=tf.int32)
-        costs = tf.TensorArray(dtype=cost.dtype, size=int(self.iter_max))
+        costs = tf.TensorArray(dtype=cost.dtype, size=0, dynamic_size=True)
```
Comment (tgregov) [inline igm/processes/iceflow/unified/optimizers/lbfgs.py]: Can you avoid having the `dynamic_size=True`? I think it might deteriorate the performance. Or maybe have this as an option?

File: igm/processes/iceflow/unified/optimizers/lbfgs.py [hunk 15]
```diff
@@ -264,14 +306,14 @@ def minimize_impl(self, inputs: tf.Tensor) -> tf.Tensor:
...
-            s, y = self._constrain_pair(s, y, theta_prev, theta_trial, mask)
+            s, y = self._constrain_pair(s, y, theta_prev, theta_trial, mask, theta_flat, grad_theta_flat)
 
             # Update memory
             s_flat_mem, y_flat_mem, idx_memory = self._update_memory(
                 s_flat_mem, y_flat_mem, idx_memory, s, y
             )
 
-            # TODO: check if this is necessary
+            # this is needed e.g. for data assimilation logging
```
Comment (tgregov) [inline igm/processes/iceflow/unified/optimizers/lbfgs.py:320]: It's not really urgent, but maybe this logging step can be combined with the `update_display' step (which could be renamed if needed)?

File: igm/processes/iceflow/unified/solver/solver.py (hunk 5 above)
Comment (tgregov) [inline igm/processes/iceflow/unified/solver/solver.py]: I would also add a condition that checks that we're in the `network` mapping mode; otherwise the input normalizers do not make sense

File: igm/processes/iceflow/unified/utils.py [hunk 16]
```diff
@@ -122,4 +122,4 @@ def _print_data_preparation_summary(
             sampler.batch_size_effective,
             prep,
             num_patches,
-        )
+        )
```
Comment (tgregov) [inline igm/processes/iceflow/unified/utils.py:125]: Check that this file is saved with the code formatter :)

File: igm/processes/iceflow/unified/mappings/network.py [hunk 17]
```diff
@@ -35,6 +33,7 @@ def get_UV_impl(self) -> Tuple[tf.Tensor, tf.Tensor]:
         U, V = Y_to_UV(self.Nz, Y)
         return U, V
 
+
```
Comment (tgregov) [inline igm/processes/iceflow/unified/mappings/network.py:36]: Check that this file is saved with the code formatter :)

File: igm/processes/iceflow/unified/bcs/no_inflow.py (hunk 3 above)
Comment (tgregov) [reply igm/processes/iceflow/unified/bcs/no_inflow.py:21]: I think this should be fine :)

File: igm/processes/iceflow/unified/bcs/no_inflow.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+        U_w = tf.minimum(U[:, :, :, 0:1], 0.0)
+        # east boundary (x=end): outward normal is +x, so outflow means U >= 0
+        U_e = tf.maximum(U[:, :, :, -1:], 0.0)
+        U = tf.concat([U_w, U[:, :, :, 1:-1], U_e], axis=3)
+
+        # south boundary (y=0): outward normal is -y, so outflow means V <= 0
+        V_s = tf.minimum(V[:, :, 0:1, :], 0.0)
+        # north boundary (y=end): outward normal is +y, so outflow means V >= 0
+        V_n = tf.maximum(V[:, :, -1:, :], 0.0)
+        V = tf.concat([V_s, V[:, :, 1:-1, :], V_n], axis=2)
```
Comment (tgregov) [inline igm/processes/iceflow/unified/bcs/no_inflow.py:34]: However, i think this assumes so kind of nodal structure in the vertical discretization. For example, the Legendre basis might create something a bit weird... So to avoid any issue i would put in the constructor of the class some checks about the vertical basis, ensuring that it's either ssa or molho or lagrange, which should all be fine with this implementation

**Review by tgregov — CHANGES_REQUESTED** [review]
Hi @shrrosier, nice work!
I mainly reviewed the stuff related to iceflow. Do not hesitate to ask me if you have questions related to my comments :)

File: igm/processes/iceflow/energy/components/sliding/laws/weertman.py (hunk 2 above)
Comment (shrrosier) [reply igm/processes/iceflow/energy/components/sliding/laws/weertman.py:140]: I have done this for coulomb... for budd in order to have the same interpretation and units I think we would need to add an additional N_{ref} term... will discuss further.

File: igm/processes/iceflow/unified/solver/solver.py (hunk 5 above)
Comment (shrrosier) [reply igm/processes/iceflow/unified/solver/solver.py]: After our discussion I slightly modified this as we agreed.

File: igm/processes/iceflow/unified/optimizers/lbfgs.py (hunk 14 above)
Comment (shrrosier) [reply igm/processes/iceflow/unified/optimizers/lbfgs.py]: this potentially avoids some crash if iter_max changes but agree that for now it's not necessary so removed it

File: igm/processes/iceflow/unified/bcs/no_inflow.py (hunk 18 above)
Comment (shrrosier) [reply igm/processes/iceflow/unified/bcs/no_inflow.py:34]: Added an error in the get_bc_args for no_inflow if an incompatible vertical basis is used.

File: igm/processes/iceflow/unified/mappings/transforms.py (hunk 4 above)
Comment (shrrosier) [reply igm/processes/iceflow/unified/mappings/transforms.py:136]: all names should now use consistent convention (no caps lock)

File: igm/processes/data_assimilation_SR/terms.py (hunk 7 above)
Comment (shrrosier) [reply igm/processes/data_assimilation_SR/terms.py]: all names should now use consistent convention (no caps lock)

File: igm/processes/iceflow/emulate/utils/normalizations.py (hunk 1 above)
Comment (shrrosier) [reply igm/processes/iceflow/emulate/utils/normalizations.py:339]: For now I kept both because I don't want to break anything by deleting the previous FixedAffine layer, but added a todo comment to remove this down the line.

File: igm/processes/iceflow/unified/optimizers/lbfgs.py (hunk 8 above)
Comment (shrrosier) [reply igm/processes/iceflow/unified/optimizers/lbfgs.py]: added these as arguments with default values.
